include/revng/Support/GraphAlgorithms.h

```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
 }
 
 template<class GraphT>
-inline llvm::SmallPtrSet<GraphT, 4>
+inline SmallSetVector<GraphT>
 nodesBetweenNew(GraphT Source,
                 GraphT Destination,
-                const llvm::SmallPtrSetImpl<GraphT> *IgnoreL = nullptr) {
+                const SmallSetVector<GraphT> *IgnoreList = nullptr) {
   return nodesBetweenNewImpl<GraphT, llvm::GraphTraits<GraphT>>(Source,
                                                                 Destination,
-                                                                IgnoreL);
+                                                                IgnoreList);
 }
 
 template<class GraphT>
-inline llvm::SmallPtrSet<GraphT, 4>
+inline SmallSetVector<GraphT>
 nodesBetweenNewReverse(GraphT Source,
                        GraphT Destination,
-                       const llvm::SmallPtrSetImpl<GraphT> *IgnoreL = nullptr) {
+                       const SmallSetVector<GraphT> *IgnoreList = nullptr) {
   using namespace llvm;
   return nodesBetweenNewImpl<GraphT, GraphTraits<Inverse<GraphT>>>(Source,
                                                                    Destination,
-                                                                   IgnoreL);
+                                                                   IgnoreList);
+}
+
+template<class NodeT>
+bool intersect(SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+  // TODO: Upon migration to LLVM 17, we should use the `llvm::set_intersection`
+  // provided in `SetOperations.h` header file.
+  llvm::SmallPtrSet<NodeT, 4> Intersection(First.begin(), First.end());
+  llvm::set_intersect(Intersection, Second);
+  return !Intersection.empty();
+}
+
+template<class NodeT>
+bool subset(SmallSetVector<NodeT> &Contained,
+            SmallSetVector<NodeT> &Containing) {
+  return llvm::set_is_subset(Contained, Containing);
+}
+
+template<class NodeT>
+bool equal(SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+  return llvm::set_is_subset(First, Second)
+         and llvm::set_is_subset(Second, First);
+}
+
+template<class NodeT>
+bool simplifyRegionsStep(llvm::SmallVector<SmallSetVector<NodeT>> &R) {
+  for (auto RegionIt1 = R.begin(); RegionIt1 != R.end(); RegionIt1++) {
+    for (auto RegionIt2 = std::next(RegionIt1); RegionIt2 != R.end();
+         RegionIt2++) {
+      bool Intersects = intersect(*RegionIt1, *RegionIt2);
+      bool IsIncluded = subset(*RegionIt1, *RegionIt2);
+      bool IsIncludedReverse = subset(*RegionIt2, *RegionIt1);
+      bool AreEquivalent = equal(*RegionIt1, *RegionIt2);
+      if (Intersects
+          and (((!IsIncluded) and (!IsIncludedReverse)) or AreEquivalent)) {
+        (*RegionIt1).insert((*RegionIt2).begin(), (*RegionIt2).end());
+        R.erase(RegionIt2);
```

Comment:
Are you sure this is safe?

If you need to iterate on a container and erase certain elements at the same time usually you do

```c++
for (auto It = V.begin(); It != V.end(); /* Nothing here */) {
  if (WantToDelete) {
    It = V.erase(It);
  } else {
    ++It;
  }
}
```

Reply:
You are right in the general case, but in this particular implementation an `erase` operation on the container is always followed by a `return` from the current function, for algorithm-related reasons. So in the end, every time we have an `erase`, we conclude our visit on the container and start again. Therefore I think that the implementation is correct.

Reply:
Right, I added a comment specifying the situation in the code now moved in `revng-c`.